WalletWasabi/WabiSabi/Crypto/CredentialIssuer.cs

```diff
@@ -196,31 +218,56 @@ public ICommitableCredentialsResponse PrepareResponse(CredentialsRequest registr
 			var areProofsValid = ProofSystem.Verify(transcript, statements, registrationRequest.Proofs);
 			if (!areProofsValid)
 			{
+				// Request was invalid, but all serial numbers were unused.
+				// Ensure nullifier set can't be clogged with invalid serial
+				// numbers. Valid serial numbers are only issued in response to
+				// valid requests, which in turn depend on valid ownership
+				// proofs and the banning mechanism, and they actually matter
+				// for double spending, so only they need to be stored in the
+				// nullifier set.
+				foreach (var serialNumber in presentedSerialNumbers)
+				{
+					SerialNumbers.Remove(serialNumber, out _);
+				}
+
 				throw new WabiSabiCryptoException(WabiSabiCryptoErrorCode.CoordinatorReceivedInvalidProofs);
 			}
 
-			// Issue credentials.
-			var credentials = requested.Select(x => IssueCredential(x.Ma, RandomNumberGenerator.GetScalar())).ToArray();
+			// After this point serial numbers are committed irrevocably, even
+			// if the request fails at a higher level, and the credentials
+			// were never revealed because `Commit` was not called. This is
+			// because any request that made it this far is formally valid, and
+			// thus any invalidity with regard to state is unambiguously a
+			// double spend attempt, and there is no point in allowing those
+			// serial numbers to be reused and the round to proceed.
+
+			// Issue the credentials, but they won't actually be returned until
+			// Commit() is called.
+			var credentials = requested.Select(x => IssueCredential(x.Ma, RandomNumberGenerator.GetScalar())).ToImmutableArray();
```

Comment:
Why `ToImmutableArray()` does it makes sense?

Reply:
because it never gets mutated